File: 0130-surrounded-regions/0130-surrounded-regions.cpp
class Solution {
public:
    void dfs(int row, int col, vector<vector<char>>& board, vector<vector<int>> &vis, int n, int m, int drow[], int dcol[]){
        vis[row][col] = 1;

        for(int i = 0; i < 4; i++){
            int nrow = row + drow[i];
            int ncol = col + dcol[i];

            if(nrow >= 0 && nrow < n && ncol >= 0 && ncol < m && vis[nrow][ncol]==0 && board[nrow][ncol] == 'O'){
                dfs(nrow, ncol, board, vis, n, m, drow, dcol);
            }
        }
    }

    void solve(vector<vector<char>>& board) {
        int n = board.size();
        int m = board[0].size();
        int drow[] = {-1,0,1,0};
        int dcol[] = {0, 1, 0, -1};
        vector<vector<int>> vis(n, vector<int>(m,0));

        for(int j = 0; j < m; j++){
            if(board[0][j] == 'O' && vis[0][j] == 0){
                dfs(0, j, board, vis, n, m, drow, dcol);
            }
            if(board[n-1][j] == 'O' && vis[n-1][j] == 0){
                dfs(n-1, j, board, vis, n, m, drow, dcol);
            }
        }

        for(int i = 0; i < n; i++){
            if(board[i][0] == 'O' && vis[i][0] == 0){
                dfs(i, 0, board, vis, n, m, drow, dcol);
            }

            if(board[i][m-1] == 'O' && vis[i][m-1] == 0){
                dfs(i, m-1, board, vis, n, m, drow, dcol);
            }
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(board[i][j] == 'O' && vis[i][j] == 0){
                    board[i][j] = 'X';
                }
            }
        }
    }
};

File: 0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def dfs(self, row, col, vis, board, drow, dcol):
        vis[row][col] = 1
        n = len(board)
        m = len(board[0])

        for i in range(4):
            nrow = row + drow[i]
            ncol = col + dcol[i]

            if nrow >= 0 and nrow < n and ncol >= 0 and ncol < m and vis[nrow][ncol] == 0 and board[nrow][ncol] == 'O':
                self.dfs(nrow, ncol, vis, board, drow, dcol)

    def solve(self, board: List[List[str]]) -> None:
        n = len(board)
        m = len(board[0])
        vis = [[0] * m for _ in range(n)]
        drow = [-1, 0, 1, 0]
        dcol = [0, 1, 0, -1]

        for i in range(m):
            if board[0][i] == 'O' and vis[0][i] == 0:
                self.dfs(0, i, vis, board, drow, dcol)
            if board[n-1][i] == 'O' and vis[n-1][i] == 0:
                self.dfs(n-1, i, vis, board, drow, dcol)

        for i in range(n):
            if board[i][0] == 'O' and vis[i][0] == 0:
                self.dfs(i, 0, vis, board, drow, dcol)
            if board[i][m-1] == 'O' and vis[i][m-1] == 0:
                self.dfs(i, m-1, vis, board, drow, dcol)
        
        for i in range(n):
            for j in range(m):
                if vis[i][j] == 0 and board[i][j] == 'O':
                    board[i][j] = 'X'

        